### PR TITLE
Eq 399 Refactoring Navigator component

### PR DESCRIPTION
**Pull request by warrenbailey**
### What is the context of this PR?

Refactoring the navigation and routing components and consolidating the user action functionality into a single place.

The navigation state object is now an actual state pattern, which maintains the current user position in the questionnaire.

The routing component was actually dealing with user actions which isn't its purpose. That should be handled elsewhere and it should only do routing (which we're yet to implement).

The questionnaire manager was also dealing with user actions making it difficult to isolate how they are processed.

The user actions are now handled by the navigator using a chain of responsibility pattern. This means adding a new user action should simply be a case of creating a new UserAction class and placing it in the correct location in the chain.

I appreciate this is a big refactor but I believe it leads us to a more elegant solution going forward. I'm open to comments as there's definitely still room for improvements.
### How to review
1. Check out this branch
2. Check all the tests still work
3. Complete a survey and check nothing is broken
- [ ] Do all commits have sensible commit messages?
- [ ] Is all new code unit tested?
- [ ] Are there integration tests if needed?
- [ ] Is there sufficient logging?
- [ ] Is there documentation or is the code self-describing?
### Who worked on the PR

@warrenbailey

### CODE REVIEW

**Comment by warrenbailey** [comment]
Adding a WIP label but i'd still value some feedback on it, once I'm happy with it I'll squash the commits ready for merge.

**Comment by weapdiv-david** [comment]
Looks good so far, although I don't fully understand it all yet.  It makes perfect sense to split the user action processing out from the data, and I'm currently looking at better ways to process the form data in view of the changes we want to make around fields/responses/questions.

Keep going!

**Comment by warrenbailey** [comment]
Thanks for the feedback @weapdiv-david, if anyone else has time to review I'd appreciate it. Otherwise I'll fix the minor quantified code issues and squash the commits ready for merging.

**Comment by warrenbailey** [comment]
Rebased and squashed the commits ready for merge. I liked to take this further but wonder if it would be better to merge this now and then make another incremental refactor, otherwise this risks getting out of date with master very quickly.

**Comment by weapdiv-david** [comment]
Happy to merge this... Any other team members wanting to review it? (clue: you should)

File: app/navigation/navigator.py [hunk 1]
```diff
@@ -1,82 +1,156 @@
...
+            return self.provide_next_location(current_location)
+        else:
+            # request the next location
+            return self.next_action.process_action(action, current_location)
 
-    # destination  "group:block:section:question:<repetition>"
-    def _valid_destination(self, destination):
-        """
-        Returns a boolean indicating whether a proposed destination is a valid one
+    def get_action(self):
```
Comment (dhilton) [inline app/navigation/navigator.py]: Does this getter always return a string? If so, is it not a property that is set rather than a method?

**Comment by dhilton** [comment]
Apart from my comment above 👍 from me.

File: app/navigation/navigator.py (hunk 1 above)
Comment (warrenbailey) [inline app/navigation/navigator.py]: Yep always returns a string. I'll change it to be a property.

**Comment by warrenbailey** [comment]
Fixing code review comments and rebasing

**Comment by dhilton** [comment]
Happy to merge this LGTM 👍

**Pull request merged**